src/param.rs

```diff
@@ -0,0 +1,228 @@
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str;
+
+use num_traits::FromPrimitive;
+
+use crate::error;
+
+/// Available param keys.
+#[derive(PartialEq, Eq, Debug, Clone, Copy, Hash, PartialOrd, Ord, FromPrimitive)]
+#[repr(u8)]
+pub enum Param {
+    /// For messages and jobs
+    File = 'f' as u8,
+    /// For Messages
+    Width = 'w' as u8,
+    /// For Messages
+    Height = 'h' as u8,
+    /// For Messages
+    Duration = 'd' as u8,
+    /// For Messages
+    MimeType = 'm' as u8,
+    /// For Messages: message is encryoted, outgoing: guarantee E2EE or the message is not send
+    GuranteeE2ee = 'c' as u8,
+    /// For Messages: decrypted with validation errors or without mutual set, if neither
+    /// 'c' nor 'e' are preset, the messages is only transport encrypted.
+    ErroneousE2ee = 'e' as u8,
+    /// For Messages: force unencrypted message, either `ForcePlaintext::AddAutocryptHeader` (1),
+    /// `ForcePlaintext::NoAutocryptHeader` (2) or 0.
+    ForcePlaintext = 'u' as u8,
+    /// For Messages
+    WantsMdn = 'r' as u8,
+    /// For Messages
+    Forwarded = 'a' as u8,
+    /// For Messages
+    Cmd = 'S' as u8,
+    /// For Messages
+    Arg = 'E' as u8,
+    /// For Messages
+    Arg2 = 'F' as u8,
+    /// For Messages
+    Arg3 = 'G' as u8,
+    /// For Messages
+    Arg4 = 'H' as u8,
+    /// For Messages
+    Error = 'L' as u8,
+    /// For Messages: space-separated list of messaged IDs of forwarded copies.
+    PrepForwards = 'P' as u8,
+    /// For Jobs
+    SetLatitude = 'l' as u8,
+    /// For Jobs
+    SetLongitude = 'n' as u8,
+    /// For Jobs
+    ServerFolder = 'Z' as u8,
+    /// For Jobs
+    ServerUid = 'z' as u8,
+    /// For Jobs
+    AlsoMove = 'M' as u8,
+    /// For Jobs: space-separated list of message recipients
+    Recipients = 'R' as u8,
+    // For Groups
+    Unpromoted = 'U' as u8,
+    // For Groups and Contacts
+    ProfileImage = 'i' as u8,
+    // For Chats
+    Selftalk = 'K' as u8,
+    // For QR
+    Auth = 's' as u8,
+    // For QR
+    GroupId = 'x' as u8,
+    // For QR
+    GroupName = 'g' as u8,
+}
+
+/// Possible values for `Param::ForcePlaintext`.
+#[derive(PartialEq, Eq, Debug, Clone, Copy, FromPrimitive)]
+#[repr(u8)]
+pub enum ForcePlaintext {
+    AddAutocryptHeader = 1,
+    NoAutocryptHeader = 2,
+}
+
+/// An object for handling key=value parameter lists.
+///
+/// The structure is serialized by calling `to_string()` on it.
+///
+/// Only for library-internal use.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct Params {
+    inner: BTreeMap<Param, String>,
+}
+
+impl fmt::Display for Params {
```

Comment:
`fmt::Display` seems a bit of a hack to use for serialisation to me...  Maybe a plain `.pack()` (to keep using the old terminology)?

Reply:
`Display` is the recommended version to implement `to_string()`. So this matches expected rust semantics.